docs/protocol/core/v3.0.rst

```diff
@@ -1501,54 +1431,49 @@ Let "+" be the string concatenation operator.
 **Discover all nodes in a hierarchy**
 
-    To discover all nodes in a hierarchy, one can call ``list("meta/")``.
-     - all keys represent either explicit group or arrays.
-     - all intermediate prefixes ending in a ``/`` are implicit groups.
+    To discover all nodes in a hierarchy, one can call
+    ``list("meta/")``. All keys represent either explicit group or
+    arrays. All intermediate prefixes ending in a ``/`` are implicit
+    groups.
 
 **Delete a group or array**
 
-    To delete an array it is necessary to
-      - delete the metadata document for the array, (meta/P.array)
-      - delete all keys which prefix have path pointing to this to this array.  (data/P/\*)
+    To delete an array at path `P`:
+      - delete the metadata document for the array, ``delete(array_meta_key(P))``
+      - delete all data keys which prefix have path pointing to this to this array,
+	``delete_prefix("data/root" + P + "/")``
 
-    To delete a implicit group.
-      - delete all arrays under this group
-      - it should be sufficient to delete all the keys starting with prefix meta/P/ and data/P/
+    To delete an implicit group at path `P`:
+      - delete all nodes under this group - it should be sufficient to
+	perform ``delete_prefix("meta/root" + P + "/")`` and
+	``delete_prefix("data/root" + P + "/")``.
 
-    To delete an explicit group.
-      - delete all arrays under this group,
-      - delete all keys with meta/P/ prefix, meta/P/groups all keys with /data/P prefix,
+    To delete an explicit group at path `P`:
+      - delete the metadata document for the group, ``delete(group_meta_key(P))``
+      - delete all nodes under this group - it should be sufficient to
+	perform ``delete_prefix("meta/root" + P + "/")`` and
+	``delete_prefix("data/root" + P + "/")``.
 
-    Note that store implementation may decide to reify implicit groups and thus
-    protocol implementation should attempt to delete the .meta/P/.group file if
-    they really wish to delete an empty implicit group.
+    Note that store implementation may decide to reify implicit groups
+    and thus protocol implementation should attempt to delete the
+    group metadata file if they really wish to delete an empty
+    implicit group. @@TODO clarify this
 
     Store implementation are also allowed to delete any implicit parent of a
     deleted implicit groups, so a protocol implementation should make sure to
-    reify a parent group if they need to keep it. For example assuming the
-    following:
-
-     >>>  z = new_dataset()
-     >>>  z.create_array('/path/to/array')
-
-     >>>  z.delete_array('/path/to/array')
-
-     This may not be sufficient to delete the group ``/path/to/``, as a store
-     implementation, and thus removing ``/path/to/`` may need an implmentation
-     to explicitly call
-
-     >>> z.delete_group('/path/to/')
+    reify a parent group if they need to keep it. @@TODO clarify this
 
-     Even if an explicit group was not explicitly created.
```

Comment:
I removed the pseudocode example and added an @@TODO here because this was not immediately obvious to me what we were trying to communicate. Suggest we revisit this in a separate PR.